proposals/1533-host-specific-tasklist.md

```diff
@@ -0,0 +1,241 @@
+
+}
+
+err = ExecuteActivity(sessionCtx, ProcessFileActivity, filename).Get(sessionCtx, nil)
+
+if err != nil {
+
+    // Session(worker) has failed or activity itself returns an error.
+
+}
+
+err = ExecuteActivity(sessionCtx, UploadFileActivity, filename).Get(sessionCtx, nil)
+
+if err != nil {
+
+    // Session(worker) has failed or activity itself returns an error.
+
+}
+
+err = CompleteSession(sessionCtx)
+
+if err != nil {
+
+    // Wrong ctx is used or failed to release session resource.
+
+}
+```
+
+#### New Worker Options
+
+There will be three new options available when creating a worker:
+
+* **EnableSessionWorker**
+
+  The flag controls whether the worker will accept activities belongs to a session.
+
+* **SessionResourceID**
+
+  An identifier of the resource that will be consumed if a session is executed on the worker. For example, if a worker has a GPU, and a session for training a machine learning model is executed on the worker, then some memory on the GPU will be consumed. In this case, the resourceID can be the identifier of the GPU.
+
+**NOTE: user need to ensure there’s only one worker using a certain resourceID.**
+
+* **MaxCurrentSessionExecutionSize**
+
+  Since a session may consume some kind of resource, user can use this option to control the maximum number of sessions running in the worker process at the same time.
+
+## Implementation
+
+### Session Model
+
+All sessions are **resource specific**. It basically means a session is always tied to some kind of resource (not the worker). The resource can be CPU, GPU, memory, file descriptors etc and we want all the activities within a session to be executed by the same worker which owns the resource. If worker dies and restarts, as long as it owns the same resource, we can try to reestablish the session (this is future work). Also user needs to ensure that resource is owned by only one worker.
+
+Note that when creating a session, user doesn’t need to specify the resource that is consumed by the session. This is because right now one worker owns only one resource, so there’s only one resource type on the worker. As a result, as long as the user specifies the correct tasklist in the context passed into the createSession API, the correct type of resource will be consumed (since there’s only one). Later, a worker can support multiple types of resources and at that time, users need to specify the resource type when creating a session.
+
+### Workflow Context
+
+When user calls the CreateSession() or CreateSessionWithHostID() API, a structure contains information about the created session will be stored in the returned context. The structure contains three two pieces of information:
```

Comment:
When user -> When a user
a structure contains -> a structure that contains

The structure contains three two pieces of information -> The structure contains three pieces of information